src/types.ts

```diff
@@ -0,0 +1,122 @@
+// Copyright 2023 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * Common configuration that will be obtained from
+ * environment variables if not explicitly provided.
+ */
+export interface OAuthEnvConfig {
+  /** The client ID provided by the authorization server. */
+  clientId: string;
+
+  /** The client secret provided by the authorization server, if using a confidential client. */
+  clientSecret: string;
+
+  /** The custom domain of the authorization server, if supported by the provider. */
+  domain?: string;
+}
+
+/**
+ * Information supplied by a provider but not directly necessary for
+ * the OAuth2 process.
+ */
+export interface OAuthProviderInfo {
+  /** The name of the provider (single PascalCase name that matches the function name and uppercase env var prefix) */
+  name: string;
+}
+
+/**
+ * Configuration supplied by the providers.
+ */
+export interface OAuthPredefinedConfig {
+  /** The URI of the authorization server's authorization endpoint. */
+  authorizationEndpointUri: string | URL;
+
+  /** The URI of the authorization server's token endpoint. */
+  tokenUri: string | URL;
+
+  /**
+   * Default scopes to request, this should be the minimal read scopes for sign-in and
+   * user identity (eg. id, email), it should also include scopes for OpenID Connect if the provider supports it.
+   */
+  scope: string[];
```

Comment:
I felt it made more sense to keep the config object flat, and move scope up, out of the 'defaults' prop, which is really an implementation detail within oauth2_client, and doesn't make a lot of sense in the config for this.

Also, I felt it would provide a better DX for a provider to supply a reasonable default scope so it's easier to get going with kv_oauth, ie. no need to go off hunting for scopes just for some basic user details.

This default scope is completely overridden if supplied by the user.